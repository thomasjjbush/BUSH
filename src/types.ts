import { ReactElement, UIEvent } from 'react';

export interface AspectRatioProps {
    children?: ReactElement | ReactElement[] | string;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '3:2' = 66.66,
    '4:3' = 75,
    '8:5' = 62.5,
    '16:9' = 56.25,
    '32:9' = 28.125,
}

export interface ButtonProps {
    ariaLabel?: string;
    children: Children;
    design?: Buttons;
    onClick: () => void;
}

export enum Buttons {
    PRIMARY,
    SECONDARY,
}

export interface CarouselProps {
    children: ReactElement[];
    color?: string;
    itemsToDisplay?: number;
    margin?: number;
    numberOfItems: number;
    startingIndex?: number;
}

type Children = ReactElement | ReactElement[] | string | string[] | (ReactElement | string)[];

export interface Client {
    logo: {
        url: string;
    };
    name: string;
    primaryColor: string;
    secondaryColor?: string;
    slug: string;
    url: string;
}

export enum ClientsActions {
    LOADED = 'clients/LOADED',
}

export interface ClientProps {
    alt?: string;
    background?: boolean;
    border?: string;
    logo: string;
    size: number;
    url?: string;
    width: number;
}

type Collection<T> = { items: T[]; total?: number };

export interface Employment {
    companyName: string;
    endDate?: string;
    location: {
        lat: string;
        lon: string;
    };
    responsibilities: string;
    startDate: string;
    title: string;
    url: string;
}

export enum EmploymentActions {
    LOADED = 'employment/LOADED',
}

export interface ErrorProps {
    button?: ReactElement;
    code: number;
}

export interface Errors {
    [key: number]: string;
}

type Gallery<T> = Collection<T> & { isLandScape: boolean };

export type GetState = () => Store;

export interface HiddenTextProps {
    children: string;
    lines: number;
}

export interface IconProps {
    color?: string;
    disabled?: boolean;
    icon: Icons | string;
    onClick?: () => void;
    size?: number;
    testID?: string;
}

export enum Icons {
    CONTRAST = 'e9d5',
    ERROR = 'ea07',
    GITHUB = 'eab0',
    INFO = 'ea0c',
    LEFT = 'ea44',
    LINKEDIN = 'eac9',
    RIGHT = 'ea42',
    WHATSAPP = 'ea93',
}

export interface ImageProps {
    alt?: string;
    height?: number;
    quality?: number;
    size: number;
    src: string;
    width?: number | string;
}

interface Media {
    description?: string;
    name: string;
    url: string;
}

export enum Platform {
    DESKTOP,
    MOBILE,
    TABLET,
}

export interface Project {
    client: Client;
    description: string;
    gallery?: Gallery<Media>;
    hero: Media;
    name: string;
    primaryTag?: Tag;
    responsibilities: Collection<Responsibility>;
    slug: string;
    tags: Collection<Tag>;
    thumbnail: Media;
    video: Media;
    year: number;
}

export enum ProjectActions {
    CLEARED = 'project/CLEARED',
    ERROR = 'project/ERROR',
    LOADED_PROJECT = 'project/LOADED',
}

export interface ProjectState {
    error?: number;
    item?: Project;
    sameClient?: Partial<Project>[];
    sameTag?: Partial<Project>[];
}

export enum ProjectsActions {
    LOADED = 'projects/LOADED',
    LOADED_MORE = 'projects/LOADED_MORE',
    LOADING_MORE = 'projects/LOADING_MORE',
}

export interface ProjectsState extends Collection<Project> {
    loading?: boolean;
}

export interface Responsibility {
    description: string;
    icon: string;
    name: string;
}

export interface ScrollerProps {
    children: Children;
    loading?: boolean;
    loadMore?: () => void;
    numberOfItems?: number;
    onScroll?: (e: UIEvent<HTMLDivElement>) => void;
    testID?: string;
    totalItems?: number;
}

export interface Store {
    clients: Client[];
    employment: Employment[];
    project: ProjectState;
    projects: ProjectsState;
    tags: Tag[];
}

export type StyledProps<T = {}> = { theme: Theme } & T;

export interface Tag {
    name: string;
    slug: string;
}

export enum TagActions {
    LOADED = 'tag/LOADED',
}

export interface TagProps {
    tags: Tag[];
}

export interface Theme {
    breakpoints: {
        mobile: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        highlight: string;
        background: string;
        border: string;
        brand: string;
        offset: string;
        text: string;
    };
    images: {
        logo: {
            pink: string;
            white: string;
        };
        loading: string;
    };
    isWebpSupported?: boolean;
}

export interface ThumbnailProps {
    client?: ReactElement;
    color: string;
    name: string;
    ratio?: AspectRatios;
    url: string;
}

declare global {
    interface Window extends Window {
        __REDUX_DEVTOOLS_EXTENSION__?: Function;
        __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: Function;
    }
}
