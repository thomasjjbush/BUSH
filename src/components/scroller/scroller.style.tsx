import styled from 'styled-components';

export const StyledScroller = styled.div`
    align-items: flex-start;
    flex: 1 1 auto;
    overflow: scroll;
    padding: 3px;
    scroll-behavior: smooth;
    z-index: 2;

    ::-webkit-scrollbar {
        display: none;
    }
`;
