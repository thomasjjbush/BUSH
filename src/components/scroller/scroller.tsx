import React, { forwardRef, ReactElement } from 'react';
import { useTheme } from 'styled-components';
import { Button, Image } from '../';
import { labels } from '../../config';
import { StyledScroller } from './scroller.style';
import { ScrollerProps as Props, Theme } from './../../types';

export const Scroller = forwardRef<HTMLDivElement, Props>(
    ({ children, loading, loadMore, numberOfItems, onScroll, testID, totalItems }: Props, ref): ReactElement => {
        const { images } = useTheme() as Theme;

        return (
            <StyledScroller data-test-id={testID || 'scroller'} onScroll={onScroll} ref={ref} tabIndex={0}>
                {children}
                {loadMore && numberOfItems < totalItems && (
                    <Button data-test-id="load-more-cta" onClick={loadMore}>
                        {labels.loadMore}
                        {loading && <Image src={images.loading} size={100} width={20} alt="loading-icon" />}
                    </Button>
                )}
            </StyledScroller>
        );
    },
);

Scroller.displayName = 'Scroller';
