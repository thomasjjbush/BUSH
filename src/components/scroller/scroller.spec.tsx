import React from 'react';
import { shallow } from 'enzyme';
import { Scroller } from './scroller';
import { ScrollerProps as Props } from '../../types';
import { StyledScroller } from './scroller.style';
import { Button, Image } from '../';

describe('Tags', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            children: <div>children</div>,
            loading: false,
            loadMore: jest.fn(),
            onScroll: jest.fn(),
        };
    });

    it('should match snapshot', () => {
        expect(shallow(<Scroller {...props} />)).toMatchSnapshot();
    });

    it('should invoke onScroll cb on scroll', () => {
        const wrapper = shallow(<Scroller {...props} />);
        wrapper.find(StyledScroller).prop('onScroll')();
        expect(props.onScroll).toHaveBeenCalledTimes(1);
    });

    it('should render button if loadMore and numberOfItems is less than totalItems', () => {
        const wrapper = shallow(<Scroller {...props} numberOfItems={1} totalItems={2} />);
        wrapper.find(Button).prop('onClick')();
        expect(props.loadMore).toHaveBeenCalledTimes(1);
    });

    it('should render button with loading image if loading', () => {
        const wrapper = shallow(<Scroller {...props} loading numberOfItems={1} totalItems={2} />);
        const loadingImage = wrapper.find(Image);
        expect(loadingImage.prop('src')).toBe('theme.images.loading');
        expect(loadingImage.prop('size')).toBe(100);
        expect(loadingImage.prop('width')).toBe(20);
    });
});
