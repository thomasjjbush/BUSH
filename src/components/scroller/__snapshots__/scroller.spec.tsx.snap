// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags should match snapshot 1`] = `
<styled.div
  data-test-id="scroller"
  onScroll={[MockFunction]}
  styles="
    align-items: flex-start;
    flex: 1 1 auto;
    overflow: scroll;
    padding: 3px;
    scroll-behavior: smooth;

    ::-webkit-scrollbar {
        display: none;
    }
"
  tabIndex={0}
>
  <div>
    children
  </div>
</styled.div>
`;
